### PR TITLE
[Stevan Gerard Gunawan] iP

### PR DESCRIPTION
**Pull request by gerardstevan**
# Doke

Doke is a simple java program which would keep track of Tasks. It would utilize user inputs to produce outputs and tasks and any tasks would be stored locally in your device after every addition or changes.

It would be able to handle Task such as:

- Events
- To-Dos
- Deadlines

Features: 

- [X] Managing Task
- [x] Keep deadline and Event date
- [ ]  Further expansion of task types and features (coming soon)

All you need to do is,

1. Pull it from [here ](https://github.com/gerardstevan/ip.git)
2. Open it in your IDE or terminal.
3. compile and run the Doke.java file.
4. let it manage your **tasks** for you :upside_down_face:

If you are doing it in terminal, you can go to the directory which contains the code and enter `javac Doke.java` before entering `java Doke` if you have java installed in your device and the program should run. 
The following is the body of the main method in Doke.java.

``` java
        Ui ui = new Ui();
        Scanner sc = new Scanner(System.in);
        Storage storage = new Storage();
        Parser parser = new Parser();
        TaskList taskList;

        ui.printOut("full path: " + Storage.DOKE_FILE.getAbsolutePath());

        ui.printOut("hello");

        taskList = new TaskList(ui, storage);
        ui.printOut("The following is your stored task: \n");
        taskList.listOut(ui);

        String str = sc.nextLine();
        while (true) {
            if (!parser.processString(taskList, storage, str, ui)) {
                break;
            }
            str = sc.nextLine();
        }

        ui.printOut("bye");
```

If you are interested in contributing to the project do keep in mind this quote:

> “Always code as if the guy who ends up maintaining your code will be a violent psychopath who knows where you live”
> -John Woods

### CODE REVIEW

File: src/main/java/doke/DialogBox.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
+        this.setAlignment(Pos.TOP_LEFT);
+        ObservableList<Node> tmp = FXCollections.observableArrayList(this.getChildren());
+        FXCollections.reverse(tmp);
+        this.getChildren().setAll(tmp);
+    }
+
+    public static DialogBox getUserDialog(Label l, ImageView iv) {
```
Comment (rui-han-crh) [inline src/main/java/doke/DialogBox.java]: Perhaps consider a more explicity naming convention? l and iv may not convey the variable well unless the reader refers to the type in the formal parameter.

File: src/main/java/doke/TaskList.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+    /**
+     * A public constructor for the TaskList class.
+     * It will try to read and create and ArrayList of type task from storage,
+     * which, if not possible would then create an empty ArrayList of type Task.
+     *
+     * @param ui
+     * @param storage
+     */
+    public TaskList(Ui ui, Storage storage) {
```
Comment (rui-han-crh) [inline src/main/java/doke/TaskList.java:29]: This constructor is slightly long and the control flow may be hard to follow due to some inverted statements.
Perhaps consider shifting the Scanner responsibilities to a new method/class?

File: src/main/java/doke/Events.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package doke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Represent the Events task. Extends the Task class.
+ */
+public class Events extends Task{
```
Comment (rui-han-crh) [inline src/main/java/doke/Events.java]: Perhaps the name of this class should be `Event` to signify a singular object
```suggestion
public class Event extends Task{
```

File: src/main/java/doke/Events.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    /**
+     * Returns a string representation of the Events object.
+     *
+     * @return a string representation of the Events object.
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (at: " + at.format(DateTimeFormatter.ofPattern("dd-MMM-yyyy")) + ")";
```
Comment (rui-han-crh) [inline src/main/java/doke/Events.java]: Perhaps consider caching the `DateTimeFormatter::ofPattern` output once during construction to reduce load of toString calls?

File: src/main/java/doke/TaskList.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        }
+    }
+
+    public void searchString(String str, Ui ui) {
+        int len = taskList.size();
+        int num = 0;
+        String message = "_________________________ \n";
+        int i = 0;
+        message = "_________________________ \n";
+        while (i < len) {
```
Comment (rui-han-crh) [inline src/main/java/doke/TaskList.java:86]: It seems like i is incremented at the end of the loop no matter what, perhaps the use of the enhanced for loop will make the purpose of this loop more explicit?

File: src/test/java/doke/TodoTest.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package doke;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TodoTest {
+
+    @Test
+    public void getType_emptyString_stringTReturned() {
```
Comment (rui-han-crh) [inline src/test/java/doke/TodoTest.java:10]: I like the naming convention of these test cases - they are very clear in their domain and expected outcomes

File: src/main/java/doke/Parser.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+     * It processes the string to see if it is a valid create and add(todo, event, and deadline),
+     * delete, mark, unmark, bye, list or otherwise, invalid command.
+     *
+     * @param taskList the taskList to be modified when needed.
+     * @param storage the storage to be modified when needed.
+     * @param string the text to be processed.
+     * @param ui to print out messages
+     * @return boolean to signify whether the doke program should end or not.
+     */
+    public boolean processString(TaskList taskList, Storage storage, String string, Ui ui) {
```
Comment (rui-han-crh) [inline src/main/java/doke/Parser.java]: This function seems to be more than 100 lines long and explicitly declares the implementations of all parsed instructions. Perhaps this does not conform well to the SLAP principle and would benefit from abstracting and splitting each process in another private method and then calling these methods to improve readability and maintainability?

File: src/main/java/doke/Storage.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * Returns a String representation of the Task suitable for the storage file.
+     *
+     * @param task task to be represented.
+     * @return a string representation of the task which can be written to the storage file.
+     */
+    public static String createWordForFile(Task task) {
+        String time = task instanceof ToDo ? "" : " | " + task.getTime();
```
Comment (rui-han-crh) [inline src/main/java/doke/Storage.java:71]: I like the use of the ternary operator here

File: src/main/java/doke/DialogBox.java (hunk 1 above)
Comment (rui-han-crh) [reply src/main/java/doke/DialogBox.java]: Oh I see, I think this is the default names from the guide 😅

File: src/main/java/doke/TaskList.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+    /**
+     * Returns a copy of the ArrayList.
+     *
+     * @return a copy of the ArrayList.
+     */
+    public ArrayList<Task> getList() {
+        ArrayList<Task> temp = this.taskList;
+        return temp;
+    }
```
Comment (rui-han-crh) [inline src/main/java/doke/TaskList.java]: There seems to be an error in this function, it mentions that it would like to return a copy of the ArrayList, but the current implementation will result in aliasing the list. Perhaps consider cloning and casting the result?

```suggestion
     * Returns a copy of the ArrayList.
     *
     * @return a copy of the ArrayList.
     */
    public ArrayList<Task> getList() {
        ArrayList<Task> temp = (ArrayList<Task>)this.taskList.clone();
        return temp;
    }
```

Although, I'm unsure this if this a documentation error. If so, maybe it would be a better choice to describe the function as `Retrieves the current TaskList`, as the word `copy` may indicate the explicit acting of copying over the objects to a new collection.

File: src/main/java/doke/TaskList.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+     * It will try to read and create and ArrayList of type task from storage,
+     * which, if not possible would then create an empty ArrayList of type Task.
+     *
+     * @param ui
+     * @param storage
+     */
+    public TaskList(Ui ui, Storage storage) {
+
+        try {
+            Scanner s = new Scanner(Storage.DOKE_FILE);
```
Comment (rui-han-crh) [inline src/main/java/doke/TaskList.java:31]: I like that you opened a scanner and consumed the `DOKE.FILE`, this is a neat deserialisation method!

**Review by rui-han-crh — COMMENTED** [review]
I like that there we clearly defined test cases with explicit output results.

A greater use of abstraction could greatly enhance the potential for readability! 👍

File: src/main/java/doke/Storage.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+
+/**
+ * a class to represents the storage.
+ */
+public class Storage {
+    protected static final File DOKE_FILE = new File(Doke.DOKE_FILE_PATH);
```
Comment (sunyuting2000) [inline src/main/java/doke/Storage.java:15]: All capitalized name suggests that it is a constant instead of a variable. Good.

File: src/main/java/doke/Storage.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * Updates the storage file using the arrayList as reference.
+     *
+     * @param arrayList containing the task to be written to the file.
+     */
+    public void updateFile(ArrayList<Task> arrayList, Ui ui) {
```
Comment (sunyuting2000) [inline src/main/java/doke/Storage.java:23]: The methods are all verbs and are all in camelCase. Good.

File: src/main/java/doke/ToDo.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
+package doke;
+
+import java.time.LocalDate;
+
+/**
+ * Represent the ToDo task. Extends the Task class.
+ */
+public class ToDo extends Task{
```
Comment (sunyuting2000) [inline src/main/java/doke/ToDo.java]: There should be a space between the Task and {, like so

```suggestion
public class ToDo extends Task {
```
Other than that nice job!

**Review by Effixion — COMMENTED** [review]
Overall there were not many coding standards that were breached. Good job!
